### PR TITLE
Gan model

### PR DESCRIPTION
**Pull request by jakgrab**
Let's try the pull request

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -1,4 +1,6 @@
+from pydoc import classname
```
Comment (drfifonz) [inline src/utils.py]: unnecessary import

File: src/models.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                nn.Conv2d(
+                    input_nc := output_nc,
+                    output_nc := output_nc * 2,
+                    4,
+                    stride=2,
+                    padding=1,
+                ),
+                nn.InstanceNorm2d(output_nc),  # try nn.BatchNorm2d()
+                nn.LeakyReLU(0.2, inplace=True),
+            )
```
Comment (drfifonz) [inline src/models.py]: self.model is overwrited in any loop iteration. Try implementation as a list and use list unpacking

File: src/models.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
+from turtle import forward
```
Comment (drfifonz) [inline src/models.py]: unnecessary import

File: src/models.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+from matplotlib.pyplot import xlabel
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
```
Comment (drfifonz) [inline src/models.py]: sort imports

File: src/models.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        convolution_block = nn.Sequential(
+            nn.ReflectionPad2d(1),
+            nn.Conv2d(in_channels, kernel_size=3),
+            nn.InstanceNorm2d(in_channels),
+            nn.ReLU(inplace=True),
+            nn.ReflectionPad2d(1),
+            nn.Conv2d(in_channels, 3),
+            nn.InstanceNorm2d(in_channels),
+        )
+        self.conv_block = nn.Sequential(*convolution_block)
```
Comment (drfifonz) [inline src/models.py]: Why u unpacking and saving exactly same items i nn.Sequential()?

Comment (drfifonz) [inline src/models.py]: Is nn.Sequential() iterrable?

File: src/models.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            # TODO if needed change conv2d and instancenorm arguments to
+            # meet our needs
+            nn.Conv2d(in_channels, 64, 7),
+            nn.InstanceNorm2d(64),
+            nn.ReLU(inplace=True),
+        )
+
+        # downsampling
+        in_channels = 64
+        out_channels = in_channels * 2
```
Comment (drfifonz) [inline src/models.py:20]: U can declare variables with walrus operator in 52 line. It's optional

File: src/models.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        for _ in range(2):
+            self.model += nn.Sequential(
+                nn.ConvTranspose2d(
+                    in_channels, out_channels, 3, stride=2, padding=1, output_padding=1
+                ),
+                nn.InstanceNorm2d(out_channels),
+                nn.ReLU(inplace=True),
+            )
+            in_channels = out_channels
+            out_channels = in_channels // 2
```
Comment (drfifonz) [inline src/models.py]: same as Discriminator class. Overwriting instead of adding new layers

File: src/models.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        for _ in range(2):
+            self.model = nn.Sequential(
+                nn.Conv2d(
+                    in_channels, out_channels, kernel_size=3, stride=2, padding=1
+                ),
+                nn.InstanceNorm2d(out_channels),
+                nn.ReLU(inplace=True),
+            )
+            in_channels = out_channels
+            out_channels = in_channels * 2
```
Comment (drfifonz) [inline src/models.py]: same as Discriminator class. Overwriting instead of adding new layers

File: src/models.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+        # temporary solution to be changed later
+        model = []
+
+        # residual blocks
+        for _ in range(res_blocks):
+            model += [ResidualBlock(in_channels)]
+
+        # adding residual blocks
+        self.model = nn.Sequential(*model)
```
Comment (drfifonz) [inline src/models.py]: right here u did it correctly. Why other ones differ?

File: src/models.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+                nn.InstanceNorm2d(out_channels),
+                nn.ReLU(inplace=True),
+            )
+            in_channels = out_channels
+            out_channels = in_channels // 2
+
+        # output layer
+        self.model = nn.Sequential(
+            nn.ReflectionPad2d(3), nn.Conv2d(64, out_channels, 7)
+        )
```
Comment (drfifonz) [inline src/models.py]: same as Discriminator class. Overwriting instead of adding new layers

File: src/models.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+            nn.Conv2d(input_nc, output_nc := 64, 4, stride=2, padding=1),
+            nn.LeakyReLU(0.2, inplace=True),
+        )
+
+        # 2nd
+        for _ in range(layers_number):
+            self.model = nn.Sequential(
+                nn.Conv2d(
+                    input_nc := output_nc,
+                    output_nc := output_nc * 2,
```
Comment (drfifonz) [inline src/models.py]: unnecessary walrus operator

File: src/models.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+# work-in-progress name
+class Shadower(nn.Module):
+    def __init__(self, in_channels, out_channels, res_blocks=9):
+        super(Shadower, self).__init__()
+
+        # initial conv layer
+        self.model = nn.Sequential(
+            nn.ReflectionPad2d(3),
+            nn.Conv2d(in_channels + 1, 64, 7),  # + mask
```
Comment (drfifonz) [inline src/models.py:80]: what does it mean? # + mask

File: src/models.py [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        # output layer
+        self.model = nn.Sequential(
+            nn.ReflectionPad2d(3), nn.Conv2d(64, out_channels, 7)
+        )
+
+    def forward(self, x):
+        return (self.model(x) + x).tanh()
+
+
+# work-in-progress name
```
Comment (drfifonz) [inline src/models.py]: Maybe GeneratorShadower() ?

File: src/models.py [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+
+        # downsampling
+        in_channels = 64
+        out_channels = in_channels * 2
+        for _ in range(2):
+            self.model = nn.Sequential(
+                nn.Conv2d(in_channels, out_channels, 3, stride=2, padding=1),
+                nn.InstanceNorm2d(out_channels),
+                nn.ReLU(inplace=True),
+            )
```
Comment (drfifonz) [inline src/models.py]: overwriting...

File: src/models.py [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+        in_channels = out_channels
+        out_channels = in_channels // 2
+
+        # output layer
+        self.model = nn.Sequential(
+            nn.ReflectionPad2d(3), nn.Conv2d(64, out_channels, 7)
+        )
```
Comment (drfifonz) [inline src/models.py]: overwriting...

File: src/models.py [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        self.model = nn.Sequential(*model)
+
+        # upsampling
+        out_channels = in_channels // 2
+        for _ in range(2):
+            self.model = nn.Sequential(
+                nn.ConvTranspose2d(in_channels, out_channels, 3, stride=2, padding=1),
+                nn.InstanceNorm2d(out_channels),
+                nn.ReLU(inplace=True),
+            )
```
Comment (drfifonz) [inline src/models.py]: overwriting...

File: src/models.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+        self.model = nn.Sequential(
+            nn.ReflectionPad2d(3), nn.Conv2d(64, out_channels, 7)
+        )
+
+    def forward(self, x, mask):
+        return (self.model(torch.cat((x, mask), 1)) + x).tanh()
+
+
+class Discriminator(nn.Module):
+    def __init__(self, input_nc, layers_number=3) -> None:
```
Comment (drfifonz) [inline src/models.py:137]: what if user provide layers_number = -2 or 0 ? Handling exceptions

File: src/models.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+        super(Discriminator, self).__init__()
+
+        # TODO implement number of layers regulation
```
Comment (drfifonz) [inline src/models.py]: if it is done why comment isn't removed

File: src/models.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+
+        # 1st layer
+        self.model = nn.Sequential(
+            nn.Conv2d(input_nc, output_nc := 64, 4, stride=2, padding=1),
+            nn.LeakyReLU(0.2, inplace=True),
+        )
+
+        # 2nd
```
Comment (drfifonz) [inline src/models.py]: it isn't exactly 2nd layer

File: src/models.py [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+        )
+        self.conv_block = nn.Sequential(*convolution_block)
+
+    def forward(self, x):
+        return x + self.conv_block(x)
+
+
+# former Generator class
+class Deshadower(nn.Module):
+    def __init__(self, in_channels, out_channels, res_blocks=9):
```
Comment (drfifonz) [inline src/models.py:7]: i suggest add layers number regulation with exception handling

File: src/models.py [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+            nn.ReflectionPad2d(3), nn.Conv2d(64, out_channels, 7)
+        )
+
+    def forward(self, x):
+        return (self.model(x) + x).tanh()
+
+
+# work-in-progress name
+class Shadower(nn.Module):
+    def __init__(self, in_channels, out_channels, res_blocks=9):
```
Comment (drfifonz) [inline src/models.py:74]: i suggest add layers number regulation with exception handling

File: src/models.py [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+# TODO class name to be changed
+# TODO push models.py again
+
+
+class ResidualBlock(nn.Module):
+    def __init__(self, in_channels):
```
Comment (drfifonz) [inline src/models.py]: define types of arguments and what method returns

File: src/models.py (hunk 21 above)
Comment (drfifonz) [inline src/models.py:74]: define types of arguments and what method returns

File: src/utils.py [hunk 23]
```diff
@@ -7,3 +9,17 @@ def mask_generator(shadow_img, shadow_free_img) -> torch.Tensor:
     generate mask image from shadow and shadow free image
     """
     pass
+
+
+def weights_init(model):
```
Comment (drfifonz) [inline src/utils.py]: define types of arguments and what method returns

File: src/dataset.py [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+def image_loader(image_path: str):
+    """
+    loading image by pillow
+    """
+    # https://github.com/python-pillow/Pillow/issues/835
+    return Image.open(image_path).convert("RGB")
+
+
+class Dataset(torch.utils.data.Dataset):
+    def __init__(self, root, loader, transforms_=None, target_transform=None):
```
Comment (drfifonz) [inline src/dataset.py:24]: define types of arguments and what method returns

File: src/dataset.py [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+
+
+def create_dataset(directory_path: str) -> list:
+    """
+    creating list of images
+    """
+    return list(sorted(os.listdir(directory_path)))
+
+
+def image_loader(image_path: str):
```
Comment (drfifonz) [inline src/dataset.py:15]: return type

File: src/dataset.py [hunk 26]
```diff
@@ -0,0 +1,56 @@
+import os
+import numpy as np
+import torch
+import torchvision.transforms as transforms
```
Comment (drfifonz) [inline src/dataset.py:4]: new line separation

File: src/models.py [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+                    padding=1,
+                ),
+                nn.InstanceNorm2d(output_nc),  # try nn.BatchNorm2d()
+                nn.LeakyReLU(0.2, inplace=True),
+            )
+
+        # classification layer
+        self.model = nn.Conv2d(
+            input_nc := output_nc,
+            output_nc := 1,
```
Comment (drfifonz) [inline src/models.py]: unnecessary walrus operator

**Review by drfifonz — COMMENTED** [review]
U can skip comments for my code. I'm going to fix that.

File: src/models.py (hunk 27 above)
Comment (drfifonz) [reply src/models.py]: my bad its essential

File: src/models.py (hunk 2 above)
Comment (jakgrab) [reply src/models.py]: Done

File: src/models.py (hunk 3 above)
Comment (jakgrab) [reply src/models.py]: Removed

File: src/models.py (hunk 4 above)
Comment (jakgrab) [reply src/models.py]: Sorted

File: src/models.py (hunk 5 above)
Comment (jakgrab) [reply src/models.py]: Solved

Comment (jakgrab) [reply src/models.py]: Solved

File: src/models.py (hunk 6 above)
Comment (jakgrab) [reply src/models.py:20]: Walrus operator applied where needed

File: src/models.py (hunk 7 above)
Comment (jakgrab) [reply src/models.py]: Taken care of

File: src/models.py (hunk 8 above)
Comment (jakgrab) [reply src/models.py]: Taken care of

File: src/models.py (hunk 9 above)
Comment (jakgrab) [reply src/models.py]: Taken care of

File: src/models.py (hunk 10 above)
Comment (jakgrab) [reply src/models.py]: Taken care of

File: src/models.py (hunk 12 above)
Comment (jakgrab) [reply src/models.py:80]: This comment was written for us to keep in mind that there's one more input channel, which is a shadow's mask.

File: src/models.py (hunk 13 above)
Comment (jakgrab) [reply src/models.py]: We'll see

File: src/models.py (hunk 14 above)
Comment (jakgrab) [reply src/models.py]: Yep

File: src/models.py (hunk 15 above)
Comment (jakgrab) [reply src/models.py]: Noticed, fixed.

File: src/models.py (hunk 16 above)
Comment (jakgrab) [reply src/models.py]: Mhm

File: src/models.py (hunk 17 above)
Comment (jakgrab) [reply src/models.py:137]: I'll take care of it in the next commit.

File: src/models.py (hunk 18 above)
Comment (jakgrab) [reply src/models.py]: Removed

File: src/models.py (hunk 20 above)
Comment (jakgrab) [reply src/models.py:7]: Will also be done in an upcoming commit.

File: src/models.py (hunk 21 above)
Comment (jakgrab) [reply src/models.py:74]: Will also be done in an upcoming commit.

File: src/models.py (hunk 12 above)
Comment (drfifonz) [reply src/models.py:80]: Explain more that comment in next commit
